New (old) pf-bootstrap-select directive

dtaylor113 opened the pull request:
## Description
This restores the a-pf v3.x `pf-select` directive -now renamed to `pf-bootstrap-select`.   The current `pf-select` uses angular-ui-bootstrap's dropdown attributes which do not support ngModel or ngOptions; and is also a component, which doesn't support being used as an attribute in HTML.
The new (old) `pf-bootstrap-select` directive uses the `bootstrap-select` plug-in which does support ngModel and ngOptions; and can be used as an attribute in the `<select> `element.  

This was asked for by MiQ which requires the use of ngModel, ngOptions, and pf(Bootstrap)Select attributes for their `<select>` elements.

NgDoc is located [here](https://rawgit.com/dtaylor113/angular-patternfly/pf-bootstrap-select-docs/docs/index.html#/api/patternfly.select.directive:pfBootstrapSelect)

Fixes #408 

![image](https://user-images.githubusercontent.com/12733153/30555694-3fb081d8-9c76-11e7-961f-8ef6d8fa522c.png)

## PR Checklist

- [x] Unit tests are included
- [x] Screenshots are attached (if there are visual changes in the UI)
- [ ] A Designer (@beanh66) is assigned as a reviewer (if there are visual changes in the UI)
- [ ] A CSS rep (@cshinn) is assigned as a reviewer (if there are visual changes in the UI)

dtaylor113 commented:
> Only thing I see is the demo looks like it works for 1 out of 3 samples perhaps needs to be updated for the other 2.

Do you mean the "Your preferred pet is Cat."...and you'd like to see  the other dropdowns represented?

jeff-phillips-18 commented on src/select/bootstrap-select.directive.js:
```diff
@@ -0,0 +1,47 @@
+angular.module('patternfly.select').directive('pfBootstrapSelect', function ($timeout) {
```
$timeout is not needed here

jeff-phillips-18 commented on test/select/bootstrap-select.spec.js:
```diff
@@ -0,0 +1,118 @@
+
+      var select = compileSelect('<select pf-bootstrap-select ng-model="modelValue" ng-options="o as o for o in options"></select>', $scope);
+
+      $timeout.flush();
```
Is this necessary? I don't see a timeout being used.

dtaylor113 replied:
Hmm...I removed the timeouts but the unit tests failed.  Could it be due to the operations being run in post-link?

jeff-phillips-18 replied:
Maybe the bootstrap code is using a timeout.

jeff-phillips-18 replied:
or it could be the $applyAsync

cdcabrera commented:
@dtaylor113 indeed if we could get the other 2 examples dropdowns functioning/represented

dtaylor113 commented:
@cdcabrera Ok, updated the ngDoc to include the other two dropdowns.

dtaylor113 commented:
Hi @cdcabrera & @jeff-phillips-18, let me know if this PR is ready to go or not.  -thanks!

jeff-phillips-18 commented:
Was waiting for the remove of the $timeout injection.

cdcabrera commented:
@dtaylor113 so it appears the rawgit link isn't updating still seeing the old example instead of something similar to...

  
![screen shot 2017-09-19 at 5 40 51 pm](https://user-images.githubusercontent.com/3761375/30616929-e931d022-9d61-11e7-910f-97c4a2e9e8c3.png)

dtaylor113 commented:
> Was waiting for the remove of the $timeout injection.

Hi, If I remove the $timeout's the unit test breaks.

jeff-phillips-18 commented:
You could remove the injected $timeout but leave them in the unit tests.

dtaylor113 commented:
Hi @cdcabrera, updated [rawgit](https://cdn.rawgit.com/dtaylor113/angular-patternfly/pf-bootstrap-select-docs/docs/index.html#/api/patternfly.select.directive:pfBootstrapSelect#example_demo).

Also fixed copy bug in pfSelect  -thanks

dtaylor113 commented:
Hi @jeff-phillips-18,  if I remove the $timeout injection, but still call `$timeout.flush()`, I get:

```
PhantomJS 2.1.1 (Linux) pf-bootstrap-select Page with pf-select directive should generate correct options from ng-options FAILED
	Error: No deferred tasks to be flushed in /home/dtaylor/Dev/repos/patternfly/angular-patternfly/node_modules/angular-mocks/angular-mocks.js (line 127)
	flush@/home/dtaylor/Dev/repos/patternfly/angular-patternfly/node_modules/angular-mocks/angular-mocks.js:127:59
	flush@/home/dtaylor/Dev/repos/patternfly/angular-patternfly/node_modules/angular-mocks/angular-mocks.js:2094:25
	/home/dtaylor/Dev/repos/patternfly/angular-patternfly/test/select/bootstrap-select.spec.js:27:21
```
If i remove the `$timeout.flush()`'s, I get:

```
PhantomJS 2.1.1 (Linux) pf-bootstrap-select Page with pf-select directive should generate correct options from ng-options FAILED
	Expected 'a' to be 'b'.
	/home/dtaylor/Dev/repos/patternfly/angular-patternfly/test/select/bootstrap-select.spec.js:39:37
PhantomJS 2.1.1 (Linux) pf-bootstrap-select Page with pf-select directive should respond to changes in ng-options FAILED
	Expected 3 to be 4.
```

dtaylor113 commented:
@jeff-phillips-18, removed $timeout injection from directive and replaced $timeout in unit test with a digest.

jeff-phillips-18 commented:
Need semantic styled commit wording.

dtaylor113 commented:
Updated commit msg, thanks @jeff-phillips-18
